lib/geckodriver.js

```diff
@@ -0,0 +1,152 @@
+// transpile:main
+import events from 'events';
+import _ from 'lodash';
+import { logger } from 'appium-support';
+import { JWProxy } from 'appium-base-driver';
+import { SubProcess } from 'teen_process';
+
+const log = logger.getLogger('Geckodriver');
+
+const DEFAULT_HOST = '127.0.0.1';
+const DEFAULT_PORT = 29146;
+const SYSTEM_GECKODRIVER = '/usr/local/bin/geckodriver';
+const DEFAULT_INIT_URL = 'https://www.mozilla.org/';
+
+class Geckodriver extends events.EventEmitter {
+  constructor (args = {}) {
+    super();
+
+    const {
+      host = DEFAULT_HOST,
+      port = DEFAULT_PORT
+    } = args;
+
+    this.proxyHost = host;
+    this.proxyPort = port;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort, base: ''});
+    this.state = Geckodriver.STATE_STOPPED;
+    this.capabilities = {};
+  }
+
+  async start (caps, emitStartingState = true) {
+    this.capabilities = _.cloneDeep(caps);
+
+    const args = [`--host=${this.proxyHost}`, `--port=${this.proxyPort}`, '-vv'];
+
+    const startDetector = (stdout) => {
+      return stdout.indexOf('Listening') !== -1;
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.geckodriver = SYSTEM_GECKODRIVER;
```

Comment:
This hardcodes that Geckodriver is suppose to be at `'/usr/local/bin/geckodriver'`, what happens with Windows user or other scenarios where this is not the case. Maybe something that should be configurable.